lib/legacy_session_encryptor.rb

```diff
@@ -1,18 +1,151 @@
+# frozen_string_literal: true
+
 class LegacySessionEncryptor
+  CIPHERTEXT_HEADER = 'v2'
   def load(value)
-    decrypted = encryptor.decrypt(value)
+    _v2, ciphertext = value.split(':')
+    decrypted = outer_decrypt(ciphertext)
+
+    session = JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
+    kms_decrypt_sensitive_paths!(session)
 
-    JSON.parse(decrypted, quirks_mode: true).with_indifferent_access
+    session
   end
 
   def dump(value)
+    value.deep_stringify_keys!
+
+    kms_encrypt_pii!(value)
+    kms_encrypt_sensitive_paths!(value, SessionEncryptor::SENSITIVE_PATHS)
+    alert_or_raise_if_contains_sensitive_keys!(value)
     plain = JSON.generate(value, quirks_mode: true)
-    encryptor.encrypt(plain)
+    alert_or_raise_if_contains_sensitive_value!(plain, value)
+    CIPHERTEXT_HEADER + ':' + outer_encrypt(plain)
+  end
+
+  def kms_encrypt(text)
+    Base64.encode64(Encryption::KmsClient.new.encrypt(text, 'context' => 'session-encryption'))
+  end
+
+  def kms_decrypt(text)
+    Encryption::KmsClient.new.decrypt(
+      Base64.decode64(text), 'context' => 'session-encryption'
+    )
+  end
+
+  def outer_encrypt(plaintext)
+    Encryption::Encryptors::LegacyAesEncryptor.new.encrypt(plaintext, session_encryption_key)
+  end
+
+  def outer_decrypt(ciphertext)
+    Encryption::Encryptors::LegacyAesEncryptor.new.decrypt(ciphertext, session_encryption_key)
   end
 
   private
 
-  def encryptor
-    Encryption::Encryptors::SessionEncryptor.new
+  # The PII bundle is stored in the user session in the 'decrypted_pii' key.
+  # The PII is decrypted with the user's password when they successfully submit it and then
+  # stored in the session.  Before saving the session, this method encrypts the PII with KMS and
+  # stores it in the 'encrypted_pii' key.
+  #
+  # The PII is not frequently needed in its KMS-decrypted state. To reduce the
+  # risks around holding plaintext PII in memory during requests, this PII is KMS-decrypted
+  # on-demand by the Pii::Cacher.
+  def kms_encrypt_pii!(session)
+    return unless session.dig('warden.user.user.session', 'decrypted_pii')
+    decrypted_pii = session['warden.user.user.session'].delete('decrypted_pii')
+    session['warden.user.user.session']['encrypted_pii'] =
+      kms_encrypt(decrypted_pii)
+    nil
+  end
+
+  # This method extracts all of the sensitive paths that exist into a
+  # separate hash.  This separate hash is then encrypted and placed in the session.
+  # We use #reduce to build the nested empty hash if needed. If Hash#bury
+  # (https://bugs.ruby-lang.org/issues/11747) existed, we could use that instead.
+  def kms_encrypt_sensitive_paths!(session, sensitive_paths)
+    sensitive_data = {}
+
+    sensitive_paths.each do |path|
+      all_but_last_key = path[0..-2]
+      last_key = path.last
```

Comment:
I like to use splats and destructuring when we can, helps us avoid hardcoding indexes

```suggestion
      *all_but_last_key, last_key = path
```